### ISSUE TITLE
Implement constant evaluation in patterns

### ISSUE THREAD

**Issue by chloestefantsova** [dart-lang/sdk#50869]
The following tests fail due to the absent constant evaluation in patterns:

* co19/LanguageFeatures/Patterns/constant_A03_t03
* co19/LanguageFeatures/Patterns/constant_A03_t06
* co19/LanguageFeatures/Patterns/map_A01_t01
* co19/LanguageFeatures/Patterns/constant_A03_t07
* co19/LanguageFeatures/Patterns/constant_A04_t01
* co19/LanguageFeatures/Patterns/constant_A04_t02

Consider the following program as an example:

```dart
String test(List value) {
  if (value case const <int>[1, 2]) {
    return "[1, 2]";
  }
  return "other";
}

main() {
  print(test(const <int>[1, 2]));
}
```

When complied to Kernel currently, only one of `const <int>[1, 2]` is represented as a constant.

```
library /*isNonNullableByDefault*/;
import self as self;
import "dart:core" as core;

static method test(core::List<dynamic> value) → core::String {
  final core::List<dynamic> #t1 = value;
  if(#t1 =={core::List::==}{(core::Object) → core::bool} core::_GrowableList::_literal2<core::int>(1, 2)) {
    {
      return "[1, 2]";
    }
  }
  return "other";
}
static method main() → dynamic {
  core::print(self::test(#C3));
}
constants  {
  #C1 = 1
  #C2 = 2
  #C3 = <core::int*>[#C1, #C2]
}
```

It prevents the equality to work for constants as expecting, resulting in failing tests. When the program shown above is run, it prints `other`.

**Comment by johnniwinther** [comment]
Related to #50629

**Comment by johnniwinther** [comment]
This seems to have been fixed together with #50629
